### PR TITLE
Improve missing mode.

### PR DESCRIPTION
**Pull request by lordIcocain**
If crafting cpu have job which in missing mode then cpu every minute will try extract missing items from storage until missing list is empty.

Also fixed that missing mode ignore available stock and "requested" too much.

Closes: https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/18009

### CODE REVIEW

File: src/main/java/appeng/me/cluster/implementations/CraftingCPUCluster.java [hunk 1]
```diff
@@ -1569,6 +1580,34 @@ public int getRemainingOperations() {
         }
     }
 
+    public void tryExtractItems() {
```
Comment (serenibyss) [inline src/main/java/appeng/me/cluster/implementations/CraftingCPUCluster.java:1583]: This is a pretty expensive check, have you tested if this causes lag on a large craft?

File: src/main/java/appeng/crafting/v2/CraftingJobV2.java [hunk 2]
```diff
@@ -108,6 +108,10 @@ public static CraftingJobV2 deserialize(World world, ByteBuf buffer) {
         return job;
     }
 
+    public CraftingMode getCraftingMode() {
```
Comment (serenibyss) [inline src/main/java/appeng/crafting/v2/CraftingJobV2.java:111]: This is an override right? So should have `@Override` annotated

File: src/main/java/appeng/me/cluster/implementations/CraftingCPUCluster.java [hunk 3]
```diff
@@ -1300,6 +1311,7 @@ public void writeToNBT(final NBTTagCompound data) {
         data.setTag("tasks", list);
 
         data.setTag("waitingFor", this.writeList(this.waitingFor));
+        data.setTag("waitingForMissing", this.writeList(this.waitingForMissing));
```
Comment (serenibyss) [inline src/main/java/appeng/me/cluster/implementations/CraftingCPUCluster.java:1314]: Why do you need a new waitingForMissing here, is waitingFor not enough information?

File: src/main/java/appeng/me/cluster/implementations/CraftingCPUCluster.java (hunk 1 above)
Comment (serenibyss) [reply src/main/java/appeng/me/cluster/implementations/CraftingCPUCluster.java:1604]: This isn't the most thorough test but considering these profiles are basically identical I think this is a reasonable test

File: src/main/java/appeng/me/cluster/implementations/CraftingCPUCluster.java [hunk 4]
```diff
@@ -1587,6 +1601,32 @@ public int getRemainingOperations() {
...
+                    if (pg != null) {
+                        IAEItemStack extractedItems = pg.getItemInventory()
+                                .extractItems(waitingForItem, Actionable.MODULATE, this.machineSrc);
+                        if (extractedItems != null) {
+                            IAEStack notInjected = injectItems(extractedItems, Actionable.MODULATE, this.machineSrc);
+                            if (notInjected != null) { // not sure if this even need, but still
+                                pg.getItemInventory()
+                                        .injectItems((IAEItemStack) notInjected, Actionable.MODULATE, this.machineSrc);
+                            }
+                        }
```
Comment (serenibyss) [inline src/main/java/appeng/me/cluster/implementations/CraftingCPUCluster.java:1621]: If there is concern about not being able to inject the full amount, i would first do simulate extract and inject to get the proper amount you can use, then do them again on modulate. This is more performance intensive, but if its needed its needed. May be worth throwing a logger call in this inner-most if-statement to see if its ever encountered in a real-world scenario

File: src/main/java/appeng/me/cluster/implementations/CraftingCPUCluster.java (hunk 3 above)
Comment (lordIcocain) [reply src/main/java/appeng/me/cluster/implementations/CraftingCPUCluster.java:1314]: More than enough. waitingFor will include all items which currently crafting. I want to separate missing mode quene from reqular quene bacause two reasons. Performance and missing mode might extract wrong item from storage, for example wrong circuit due to weird circuit oredict and craft will stuck.

File: src/main/java/appeng/me/cluster/implementations/CraftingCPUCluster.java (hunk 4 above)
Comment (lordIcocain) [reply src/main/java/appeng/me/cluster/implementations/CraftingCPUCluster.java]: i just tested, cpu doesn't have any byte limitations for storage, so storage overflow shouldn't happen at all. And it request exactly how much cpu need. I added log which will notify players in chat, i guess after testing "overflow" part can be fully removed.

File: src/main/java/appeng/crafting/v2/CraftingJobV2.java (hunk 2 above)
Comment (lordIcocain) [reply src/main/java/appeng/crafting/v2/CraftingJobV2.java:112]: done.

File: src/main/java/appeng/me/cluster/implementations/CraftingCPUCluster.java (hunk 3 above)
Comment (serenibyss) [reply src/main/java/appeng/me/cluster/implementations/CraftingCPUCluster.java:1314]: Ah I see, I misunderstood what this nbt data actually stored, thanks for clearing it up

File: src/main/java/appeng/me/cluster/implementations/CraftingCPUCluster.java [hunk 5]
```diff
@@ -1587,6 +1601,36 @@ public int getRemainingOperations() {
...
+                if (grid != null) {
+                    final IStorageGrid pg = grid.getCache(IStorageGrid.class);
+                    if (pg != null) {
+                        IAEItemStack extractedItems = pg.getItemInventory()
+                                .extractItems(waitingForItem, Actionable.MODULATE, this.machineSrc);
+                        if (extractedItems != null) {
+                            IAEStack notInjected = injectItems(extractedItems, Actionable.MODULATE, this.machineSrc);
+                            if (notInjected != null) { // not sure if this even need, but still
+                                for (String playerName : MinecraftServer.getServer().getAllUsernames()) {
+                                    getPlayerByName(playerName).addChatMessage(
```
Comment (serenibyss) [inline src/main/java/appeng/me/cluster/implementations/CraftingCPUCluster.java:1619]: Console log is probably fine here, though I appreciate the funny message 😄

Comment (lordIcocain) [reply src/main/java/appeng/me/cluster/implementations/CraftingCPUCluster.java]: alright, done

**Review by serenibyss — APPROVED** [review]

**Pull request merged**